merkledag/merkledag.go

```diff
@@ -184,6 +184,9 @@ func (ds *dagService) GetNodes(ctx context.Context, keys []key.Key) []NodeGetter
 			select {
 			case blk, ok := <-blkchan:
 				if !ok {
+					for _, p := range promises {
+						p.Fail(ErrNotFound)
+					}
```

Comment:
Oh, I see! `!ok` is true when the channel has been closed. Yes, that makes good sense. :)

I think I see a nasty edge case here though: consider the scenario where some NodeGetters have their Node successfully sent to their `recv` channel, but nobody has called `Get()` on them yet. The above code executes, sending down `ErrNotFound` to the `err` channel of **all** NodeGetters -- even the ones that previously succeeded. So now you have some NodeGetters with data in both the `recv` channel **and** the `err` channel. However, [`select` chooses at random](https://golang.org/ref/spec#Select_statements) when there are multiple `case`s that can run. This means there's a chance that calling `Get()` on a good NodeGetter will yield an error.

This is a long winded say of saying "we should probably only fail the NodeGetters that actually failed rather than all of them".